rcore/appmanager.c

```diff
@@ -115,39 +116,80 @@ static App *_appmanager_create_app(char *name, uint8_t type, void *entry_point,
     app->type = type;
     app->header = NULL;
     app->next = NULL;
-    app->slot_id = slot_id;
+    app->app_file = app_file;
+    app->resource_file = resource_file;
     app->is_internal = is_internal;
     
     return app;
 }
 
 
+struct appdb
+{
+    uint32_t paged_incr;  //0x58F6AExx
+    uint16_t app_cache_id; //  B5 3E trek v2 also appears in app cache
+    uint16_t unk;  // 0x0FC13Exx
+    uint32_t application_id; // in app as @000037f
+    Uuid app_uuid;  // 16 bytes
+    uint32_t unk_13;
+    uint32_t unk_17;
+    uint16_t sdk_version;
+    uint16_t app_version;
+    uint16_t zeros;
+    uint8_t app_name[32];
+    uint8_t unk_arr_company[32];  // always blank
+    uint8_t unk_arr[32]; // always blank
+} __attribute__((__packed__));
+
+
 /*
  * Load the list of apps and faces from flash
  * The app manifest is a list of all known applications we found in flash
- * We scan all block regions and look for app signatures.
- * TODO The real app likely does nothing quite so crude. We need to find the app table!
+ * We load all entries from `appdb` file.
+ * TODO: appdb seems to have duplicates (in my case), maybe we should use `pmap`, but it was missing some entries for me
  */
 void _appmanager_flash_load_app_manifest(void)
 {
+    File appdb_file;
+    if (!fs_find_file(&appdb_file, "appdb"))
+    {
+        KERN_LOG("app", APP_LOG_LEVEL_ERROR, "APPDB file not found");
+        return;
+    }
+
+    char buffer[14];
+    struct appdb appdb;
     ApplicationHeader header;
-    char buf[16];
-    // super cheesy
-    // 8 app slots
-    for(uint8_t i = 0; i < 32; i++)
+
+    // skipping 8 bytes for appdb file header
+    for (uint32_t offset = 8; offset < appdb_file.size; offset += sizeof(struct appdb))
     {
-        flash_load_app_header(i, &header);
-        
+        flash_read_file_bytes(&appdb_file, offset, (void*) &appdb, sizeof(struct appdb));
+        if (appdb.application_id == 0xFFFFFFFFu)
+            break;
+
+        File app_file;
+        snprintf(buffer, 14, "@%08d/app", appdb.application_id);
+        if (!fs_find_file(&app_file, buffer))
+            continue;
+
+        File res_file;
```

Comment:
could potentially save some stack memory by having only one file open at a time and loading these sequentially

Reply:
those are pretty small structs, and I want to store both of them in `App`